src/main/java/jchess/ruleengine/KingRulesProcessor.java

```diff
@@ -0,0 +1,196 @@
+				oRule.getRuleData().setMaxRecurrenceCount(Integer.MAX_VALUE);
+				oRule.getRuleData().setFile(fDirection);
+				oRule.getRuleData().setRank(Rank.SAME);
+				oRule.getRuleData().setFamily(Family.IGNORE);
+				oRule.getRuleData().setManoeuvreStrategy(Manoeuvre.FILE_AND_RANK);
+				oRule.getRuleData().setCustomName("MOVE[KING_CASTLING]");
+				((IRuleAgent)oRule).reset();
+				
+				for (IPositionAgent moveCandidateEntry : pathToRook) {
+					if(isFieldEndangered(moveCandidateEntry, oRuleProcessor, oBoard, oCurrentPlayer)) {
+						return;
+					}
+					// if field is endangered then quit the function
+				}
+				
+				// if we got this far the path is safe and we can add the move candidate to the map of possible move candidates.
+				IMoveCandidate oCastlingMoveKingCandidate = new MoveCandidate((IRuleAgent)oRule, oPiece, oPiece.getPosition(), possibleKingTarget);
+				IMoveCandidate oCastlingMoveRookCandidate = new MoveCandidate(null, posNextField.getPiece(), posNextField, possibleRookTarget);
+				
+				oCastlingMoveKingCandidate.addSecondaryMove(oCastlingMoveRookCandidate);
+				mpCandidateMovePositions.put(possibleKingTarget.getName(), oCastlingMoveKingCandidate);
+				
+				break;				
+			
+			}
+			else {
+				// no rook was found.
+				break;
+			}
+		}
+	}
+	
+	/**
+	 * This method executes the castling move.
+	 * 
+	 * @param IBoardAgent
+	 * @param IMoveCandidate
+	 * @return IBoardActivity
+	 */
+	public static IBoardActivity tryExecuteCastlingRule(IBoardAgent oBoard, IMoveCandidate oMoveCandidate) {
+		// move the king
+		IBoardActivity oActivity = new BoardActivity(oMoveCandidate);
+		
+		IPieceAgent oPieceLinkedToCurrentPosition = oMoveCandidate.getSourcePosition().getPiece();
+		IPositionAgent oCurrentPosition = oMoveCandidate.getSourcePosition();
+		IPositionAgent oNewPosition = oMoveCandidate.getCandidatePosition();
+		
+		oPieceLinkedToCurrentPosition.setPosition(oNewPosition);
+		oNewPosition.setPiece(oPieceLinkedToCurrentPosition);
+		oCurrentPosition.setPiece(null);
+		
+		oPieceLinkedToCurrentPosition.enqueuePositionHistory(oCurrentPosition);
+		
+		oActivity.addPriorMoveEntry(oCurrentPosition, oPieceLinkedToCurrentPosition);
+		oActivity.addPriorMoveEntry(oNewPosition, null);
+		oActivity.addPostMoveEntry(oCurrentPosition, null);
+		oActivity.addPostMoveEntry(oNewPosition, oPieceLinkedToCurrentPosition);		
+		
+		// move the rook
+		for (IMoveCandidate secondaryMoveCandidate : oMoveCandidate.getSecondaryMoves()) {
+			secondaryMoveCandidate.getPieceToMove().setPosition(secondaryMoveCandidate.getCandidatePosition());
+			secondaryMoveCandidate.getCandidatePosition().setPiece(secondaryMoveCandidate.getPieceToMove());
+			secondaryMoveCandidate.getSourcePosition().setPiece(null);
+			
+			oPieceLinkedToCurrentPosition.enqueuePositionHistory(oCurrentPosition);
+			oActivity.addPriorMoveEntry(secondaryMoveCandidate.getSourcePosition(), secondaryMoveCandidate.getSourcePosition().getPiece());
```

Comment:
By this time secondaryMoveCandidate.getSourcePosition().getPiece() is pointing to new position, which you did update in the above lines...

please make local variable even you can copypaste the main code and you will just need to change the variable names.

Reply:
E.g. (Following code is not complete..and names are poorly choosen.. please take it for ref only.)

IPieceAgent oPieceLinkedToCurrentPosition_secondaryMoveCandidate = secondaryMoveCandidate.getSourcePosition().getPiece();
		IPositionAgent oCurrentPosition_secondaryMoveCandidate  = secondaryMoveCandidate.getSourcePosition();
		IPositionAgent oNewPosition_secondaryMoveCandidate = secondaryMoveCandidate.getCandidatePosition();

		oPieceLinkedToCurrentPosition_secondaryMoveCandidate.setPosition(oNewPosition_secondaryMoveCandidate);
		oNewPosition_secondaryMoveCandidate.setPiece(oPieceLinkedToCurrentPosition_secondaryMoveCandidate);
		oCurrentPosition_secondaryMoveCandidate.setPiece(null);

		oPieceLinkedToCurrentPosition_secondaryMoveCandidate.enqueuePositionHistory(oCurrentPosition_secondaryMoveCandidate);

		oActivity.addPriorMoveEntry(oCurrentPosition_secondaryMoveCandidate, oPieceLinkedToCurrentPosition_secondaryMoveCandidate);
		oActivity.addPriorMoveEntry(oNewPosition_secondaryMoveCandidate, null);
		oActivity.addPostMoveEntry(oCurrentPosition_secondaryMoveCandidate, null);
		oActivity.addPostMoveEntry(oNewPosition_secondaryMoveCandidate, oPieceLinkedToCurrentPosition_secondaryMoveCandidate);